nntrainer/src/databuffer.cpp

```diff
@@ -100,171 +101,113 @@ bool DataBuffer::init(int mini_batch, unsigned int train_bufsize, unsigned int v
+      rest_size = &rest_test;
+      cur_size = &cur_test_bufsize;
+      running = &test_running;
+      data = &testData;
+      datalabel = &testDataLabel;
+      break;
+    default:
+      break;
+  }
 
-      for (unsigned int i = 0; i < max_train; ++i) {
-        mark[i] = i;
-      }
+  unsigned int I;
+  std::vector<unsigned int> mark;
+  mark.resize(max_size);
+  file.seekg(0, std::ios_base::end);
+  int64_t file_length = file.tellg();
 
-      unsigned int I;
-      while (train_running && mark.size() != 0) {
-        if (train_bufsize - cur_train_bufsize > 0 && rest_train > 0) {
-          data_lock.lock();
-          std::vector<float> vec;
-          std::vector<float> veclabel;
+  for (unsigned int i = 0; i < max_size; ++i) {
+    mark[i] = i;
+  }
 
-          unsigned int id = rangeRandom(0, mark.size() - 1);
-          I = mark[id];
-          if (I > max_test)
-            throw std::runtime_error("Error: Test case id cannot exceed maximum number of test");
+  while ((*running) && mark.size() != 0) {
+    if (buf_size - (*cur_size) > 0 && (*rest_size) > 0) {
+      data_lock.lock();
+      std::vector<float> vec;
+      std::vector<float> veclabel;
 
-          mark.erase(mark.begin() + id);
+      unsigned int id = rangeRandom(0, mark.size() - 1);
+      I = mark[id];
+      if (I > max_size)
+        ml_loge("Error: Test case id cannot exceed maximum number of test");
 
-          int64_t position = (I * input_size + I * class_num) * sizeof(float);
+      mark.erase(mark.begin() + id);
+      int64_t position = (I * input_size + I * class_num) * sizeof(float);
 
-          if (position > file_length)
-            throw std::runtime_error("Error: Cannot exceed max file size");
+      if (position > file_length)
+        ml_loge("Error: Cannot exceed max file size");
 
-          file.seekg(position, std::ios::beg);
+      file.seekg(position, std::ios::beg);
 
-          for (unsigned int j = 0; j < input_size; ++j) {
-            float data;
-            file.read((char *)&data, sizeof(float));
-            vec.push_back(data);
-          }
-          trainData.push_back(vec);
-          for (unsigned int j = 0; j < class_num; ++j) {
-            float data;
-            file.read((char *)&data, sizeof(float));
-            veclabel.push_back(data);
-          }
-          trainDataLabel.push_back(veclabel);
-          rest_train--;
-          cur_train_bufsize++;
-          data_lock.unlock();
-        }
-        if (train_bufsize == cur_train_bufsize) {
+      for (unsigned int j = 0; j < input_size; ++j) {
+        float d;
+        file.read((char *)&d, sizeof(float));
+        vec.push_back(d);
+      }
+      data->push_back(vec);
+      for (unsigned int j = 0; j < class_num; ++j) {
+        float d;
+        file.read((char *)&d, sizeof(float));
+        veclabel.push_back(d);
+      }
+      datalabel->push_back(veclabel);
+      (*rest_size)--;
+      (*cur_size)++;
+      data_lock.unlock();
```

Comment:
If this ```data_lock``` is supposed to protect ```data``` vector, reduce the critical section; i.e., protect from Line178 to Line184.  And even Line179-183 appears not to require protection from data_lock.

Larger critical sections will deteriorate the performance.

Reply:
Fixed. Thanks.